pybaseball/datahelpers/postprocessing.py

```diff
@@ -18,24 +18,58 @@
     (re.compile(r'^\d{4}-\d{1,2}-\d{1,2}T\d{2}:\d{2}:\d{2}.\d{1,6}Z$'), '%Y-%m-%dT%H:%M:%S.%fZ'),
 ]
 
+
 def try_parse_dataframe(
     data: pd.DataFrame,
     null_replacement: Union[str, int, float, datetime] = np.nan,
     known_percentages: List[str] = []
 ) -> pd.DataFrame:
-    values = [
-        {column: try_parse(data[column][row_i], column) for column in data.columns}
-        for row_i in range(len(data)) 
-    ]
+    data_copy = data.copy()
+    data_copy = coalesce_nulls(data_copy, null_replacement)
+    data_copy = data_copy.apply(pd.to_numeric, errors='ignore', downcast='integer').convert_dtypes(convert_string=False)
```

Comment:
Is downcast correct here? Can you explain the particular chain here?

Reply:
`downcast='integer'` should try to cast it to an integer if possible, but will settle for a float if that's what the data is.

https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.to_numeric.html

> If not None, and if the data has been successfully cast to a numerical dtype (or if the data was numeric to begin with), downcast that resulting data to the smallest numerical dtype possible according to the following rules:
> 
> ‘integer’ or ‘signed’: smallest signed int dtype (min.: np.int8)
> 
> ‘unsigned’: smallest unsigned int dtype (min.: np.uint8)
> 
> ‘float’: smallest float dtype (min.: np.float32)
> 
> As this behaviour is separate from the core conversion to numeric values, any errors raised during the downcasting will be surfaced regardless of the value of the ‘errors’ input.
> 
> In addition, downcasting will only occur if the size of the resulting data’s dtype is strictly larger than the dtype it is to be cast to, so if none of the dtypes checked satisfy that specification, no downcasting will be performed on the data.

Reply:
Is an int8 good though? A single season's G should be an int8, a career won't be though

Reply:
It just tries to go to the smallest int type, if the data is larger than that it will use an appropriate int type. For example I tested it on '4294967296' (2^32) and got the correct int back:

```python
            raw_data = pd.DataFrame(
            [
                ['1', 'TBR', '2019-01-01', '0.5', '40%', 8, '1048576'],
                ['2', 'NYY', '2019-02-01', '0.6', '20%', 'null', '4294967296']
            ],
            columns=['id', 'team', 'dt', 'rate', 'chance', 'wins', 'long_number']
        )
        processed_data = postprocessing.try_parse_dataframe(raw_data)
        print(processed_data)
```

```
    id team  dt           rate    chance  wins  long_number
0   1  TBR   2019-01-01   0.5     0.4     8     1048576
1   2  NYY   2019-02-01   0.6     0.2     <NA>  4294967296
```

Reply:
That wasn't my concern. My concern is does it automatically upcast instead of overflowing?

Reply:
which it does

Reply:
@schorrm in your comment above, "it does" means it upcasts or it overflows?

seems safer to just use `int64`; example where it could be a problem, 
```
>>> df1 = pd.DataFrame({'G': [127]})
>>> df2 = pd.DataFrame({'G': [127]})
>>> df1 = df1.apply(pd.to_numeric, errors='ignore', downcast='integer')
>>> df2 = df2.apply(pd.to_numeric, errors='ignore', downcast='integer')
>>> df1.G+df2.G
0   -2
```

Reply:
Good catch. I tried sum, I didn't get this one. No downcasting then

Reply:
Well if that's the case we're back to some sort of schema type solution to say which columns should be which types.

Reply:
Well we're never gonna overflow a 64. Convert to int as possible, but not i8 I guess?

Reply:
I don't think you can choose. If you say to "integer" (or "signed or "unsigned") it uses the smallest possible one it can. Only other option is float.

After you pointed this out I did an experiment where as long as you use pandas' sum function it will upcast it:

```python
>>> import pandas as pd
>>> df = pd.DataFrame({'G': ['127', '127', '127']})
>>> df = df.apply(pd.to_numeric, errors='ignore', downcast='integer')
>>> df['G'].sum()
381
```

I'll also play around with Pandera and see how it may help.

Reply:
Why do you need to downcast at all though?

Reply:
This is the only way in pd.to_numeric to get it to coerce integers. Otherwise you end up with a bunch of XXX.0 floats for int columns.

Reply:
Why aren't we using the `convert_dtypes` here?

Reply:
We are.

Reply:
@schorrm I think I see what you are saying now. I think I have a leaner solution incoming now.